packages/compass-query-bar/src/components/query-history-button-popover.tsx

```diff
@@ -0,0 +1,136 @@
+import React, { useCallback, useRef, useState } from 'react';
+import {
+  Icon,
+  Popover,
+  css,
+  cx,
+  focusRingStyles,
+  focusRingVisibleStyles,
+  spacing,
+  useOnClickOutside,
+} from '@mongodb-js/compass-components';
+import { createLoggerAndTelemetry } from '@mongodb-js/compass-logging';
+import type AppRegistry from 'hadron-app-registry';
+
+const { track } = createLoggerAndTelemetry('COMPASS-QUERY-BAR-UI');
+
+const openQueryHistoryStyles = cx(
+  css({
+    border: 'none',
+    backgroundColor: 'transparent',
+    display: 'inline-flex',
+    alignItems: 'center',
+    padding: spacing[2] - 2, // -2px for border.
+    '&:hover': {
+      cursor: 'pointer',
+    },
+    '&:focus': focusRingVisibleStyles,
+  }),
+  focusRingStyles
+);
+
+const queryHistoryContainerStyles = css({
+  display: 'flex',
+  height: '100%',
+});
+
+const queryHistoryPopoverStyles = css({
+  // We want the popover to open almost to the shell at the bottom of Compass.
+  maxHeight: 'calc(100vh - 270px)',
+  display: 'flex',
+  marginLeft: -spacing[4], // Align to the left of the query bar.
+});
+
+type QueryHistoryProps = {
+  globalAppRegistry: AppRegistry;
+  localAppRegistry: AppRegistry;
+};
+
+export const QueryHistoryButtonPopover: React.FunctionComponent<
+  QueryHistoryProps
+> = ({ globalAppRegistry, localAppRegistry }) => {
+  const queryHistoryRef = useRef<{
+    component: React.ComponentType<any>;
+    store: any; // Query history store is not currently typed.
+    actions: any; // Query history actions are not typed.
+  }>({
+    component: globalAppRegistry.getRole('Query.QueryHistory')![0].component,
```

Comment:
Really don't think we should be non-null asserting there, especially as this not throwing relies on a completely different prop in a completely different component being passed correctly, it think you should rather do an optional chaining here instead and handle a possibly undefined component correctly